engine/execution/ingestion/stop_control.go

```diff
@@ -0,0 +1,235 @@
+		return true
+	}
+
+	if s.state == StopControlPaused {
+		return false
+	}
+
+	// skips blocks at or above requested stop height
+	if b.Height >= s.height {
+		s.log.Warn().Msgf("Skipping execution of %s at height %d because stop has been requested at height %d", b.ID(), b.Height, s.height)
+		s.log.Debug().Int8("previous_state", int8(s.state)).Int8("new_state", int8(StopControlCommenced)).Msg("StopControl state transition")
+		s.state = StopControlCommenced // if block was skipped, move into commenced state
+		return false
+	}
+
+	return true
+}
+
+// blockFinalized should be called when a block is marked as finalized
+func (s *StopControl) blockFinalized(ctx context.Context, execState state.ReadOnlyExecutionState, h *flow.Header) {
+
+	s.Lock()
+	defer s.Unlock()
+
+	if s.state == StopControlOff || s.state == StopControlPaused {
+		return
+	}
+
+	// Once finalization reached stop height we can be sure no other fork will be valid at this height,
+	// if this block's parent has been executed, we are safe to stop or crash.
+	// This will happen during normal execution, where blocks are executed before they are finalized.
+	// However, it is possible that EN block computation progress can fall behind. In this case,
+	// we want to crash only after the execution reached the stop height.
+	if h.Height == s.height {
+
+		executed, err := state.IsBlockExecuted(ctx, execState, h.ParentID)
+		if err != nil {
+			// any error here would indicate unexpected storage error, so we crash the node
+			s.log.Fatal().Err(err).Str("block_id", h.ID().String()).Msg("failed to check if the block has been executed")
+			return
+		}
+
+		if executed {
+			s.stopExecution()
+		} else {
+			s.stopAfterExecuting = h.ParentID
+		}
+
+	}
+
+}
+
+// blockExecuted should be called after a block has finished execution
+func (s *StopControl) blockExecuted(h *flow.Header) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.state == StopControlPaused || s.state == StopControlOff {
+		return
+	}
+
+	if s.stopAfterExecuting == h.ID() {
+		// double check. Even if requested stop height has been changed multiple times,
+		// as long as it matches this block we are safe to terminate
+
+		if h.Height == s.height-1 {
+			s.stopExecution()
+		}
+	}
+}
+
+func (s *StopControl) stopExecution() {
+	if s.crash {
+		s.log.Fatal().Msgf("Crashing as finalization reached requested stop height %d", s.height)
+	} else {
+		s.log.Debug().Int8("previous_state", int8(s.state)).Int8("new_state", int8(StopControlPaused)).Msg("StopControl state transition")
+		s.state = StopControlPaused
+		s.log.Warn().Msgf("Pausing execution as finalization reached requested stop height %d", s.height)
+	}
+}
+
+// executingBlockHeight should be called while execution of height starts, used for internal tracking of the minimum
+// possible value of height
+func (s *StopControl) executingBlockHeight(height uint64) {
+	if s.state == StopControlPaused {
+		return
+	}
+
+	if height > s.lastExecutingHeight {
```

Comment:
Consider renaming to `highest`.
```suggestion
	// updating the highest executing height, which will be used to reject setting stop height that 
	// is too low. 
	if height > s.highestExecutingHeight {
```